File: 1_Two_Sum/1.cpp
#include <iostream>
#include <vector>
#include <map>

//hash table problem
//unordered_map find O(1) vs map find O(log n)
using namespace std;

class Solution {
public:
	vector<int> twoSum(vector<int> &numbers, int target) {
		int len = numbers.size();
		vector<int> v1;
		unordered_map<int,int> m;
		unordered_map<int,int>::iterator mit;
		for (int i = 0; i < len; i++) {
			mit = m.find(target - numbers[i]);
			if(mit == m.end()) 
			    m.insert(pair<int,int>(numbers[i],i+1));
			else {
				v1.push_back(mit->second);
				v1.push_back(i+1);
				break;
			}
		}
		return v1;
	}
};

int main () {
	Solution soln;
	vector<int> v;
	v.push_back(2);
	v.push_back(7);
	v.push_back(9);
	v.push_back(11);
	cout << soln.twoSum(v,9)[0]<<" "<<soln.twoSum(v,9)[1] <<endl;
	return 0;
}

File: 1_Two_Sum/1.java
public class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer,Integer> a = new HashMap<Integer,Integer>();
        int[] ret = new int[2];
        for(int i = 0; i < nums.length;i++){
            if(a.containsKey(target-nums[i])){
                ret[0] = a.get(target-nums[i]);
                ret[1] = i+1;
                return ret;
            }else
                a.put(nums[i],i+1);
        }
        return ret;
    }
}